src/core/models.py

```diff
@@ -31,36 +31,21 @@ def __init__(
         self.data_interface = data_interface
         self.component = component
         self.substrate = substrate
-
-    @property
-    def relation_data(self) -> MutableMapping[str, str]:
-        """The raw relation data."""
-        if not self.relation or not self.data_interface:
-            return {}
-
-        my_data = {}
-        if my_data_dict := self.data_interface.fetch_my_relation_data([self.relation.id]):
-            my_data = my_data_dict.get(self.relation.id, {})
-
-        try:
-            other_data = self.data_interface.fetch_relation_data([self.relation.id])[
-                self.relation.id
-            ]
-        except NotImplementedError:
-            other_data = {}
-        return my_data | other_data
+        if self.relation and self.data_interface:
+            self.relation_data = self.data_interface.as_dict(self.relation.id)
 
     def update(self, items: dict[str, str]) -> None:
         """Writes to relation_data."""
-        if not self.relation or not self.data_interface:
+        if not hasattr(self, "relation_data"):
             return
 
         delete_fields = [key for key in items if not items[key]]
-        update_fields = {k: items[k] for k in items if k not in delete_fields}
-        if update_fields:
-            self.data_interface.update_relation_data(self.relation.id, update_fields)
-        if delete_fields:
-            self.data_interface.delete_relation_data(self.relation.id, delete_fields)
+        update_content = {k: items[k] for k in items if k not in delete_fields}
+
+        self.relation_data.update(update_content)
+
+        for field in delete_fields:
+            del self.relation_data[field]
```

Comment:
...or we could just simply get rid of the whole update() method and use the "`dict`" directly.
Like databag ;-)

NOTE: With the current implementation I'm demonstrating above, we do raise an `AttributeError` on `StateBase` objects when no relation was passed, yet we wanna access relation data.

To me that makes sense, but we could think of alternatives if it seems bad for others.

Also, I think we could just turn `relation` into `relation_id`. That seems all possible (I checked references in the code) however if we wanna instantiate `StateBase` "children" objects with NO relation yet, then we've still gotta handle the "no relation ID/data" case.